### DISCUSSION TITLE
Should the extension work on html files for text content?

### DISCUSSION THREAD

**Discussion by ra-jeev** [ra-jeev/write-assist-ai#5] [Ideas]
Taken from the discussion on issue #1 

> But what can be useful is a a work in HTML files, assuming there are some blocks of plain text.
> Not sure how good models respond when output have to be in markdown format or html - what I tried manually didn't look very promising, but may be it is a matter of prompt engineering.

**Comment by ra-jeev** [comment]
Getting the output in markdown or html is doable. For adding support for LaTeX files, I took some online snippet of the format for testing, then I thought if the user selects the syntax elements as well while rephrasing then the syntax should be maintained in the response also. See the below video which illustrates what I'm talking about

https://github.com/ra-jeev/write-assist-ai/assets/17711017/32b73cf6-a368-48a5-a406-8c73feeeb9c5

As you can see the syntax (\item), though not perfect, was maintained (the same works with markdown as well). I believe this is what you mean to say @vsukhoml?

I had to add a little bit to the prompt to achieve this (sometimes it still won't follow the command): `If the text contains any special syntax then strictly follow the same syntax, e.g. for markdown return markdown, for latex return latex etc. Do not return markdown for latex, and vice versa.`

**Comment by vsukhoml** [comment]
Oh, this is cool! Yes, having such prompt and may be even making it configurable is really useful.  Sometimes you need to add more context on what the document you are working on is about as it may be not clear from selected text.

**Comment by ra-jeev** [comment]
@vsukhoml I've released the v0.3.0 just now, which makes this extension almost entirely configurable. You can modify the actions, their prompts, add new ones etc. It uses the `v1/chat/completions` endpoint now, so we can try the newer chat models also.

Please do check when you get the time, and let me know your thoughts.

Cheers :-)

**Comment by vsukhoml** [comment]
Thank you so much! Just updated and it works as expected - I tried different models and switched to gpt4-turbo-preview as a cheaper option, so will see how it goes. Will explore more the power of custom prompts - I have some ideas on domain specific prompts, but they need experiments to validate. I also suspect it can be applied to code & markup generation from requests in a more convenient way than Copilot & similar tools focused just on code.

**Comment by ra-jeev** [comment]
Please do let me know if something can be changed in the extension itself.

> I also suspect it can be applied to code & markup generation from requests in a more convenient way

Will be great to see an example so that I can understand more.

**Reply by vsukhoml** [reply]
With proper system prompt and user prompt it is possible to set something like:
`system:` "You are an experienced software engineer helping user to write code. Help with user questions to explain code, generate code. Make sure to provide step by step reasoning".
`user:` "Write the code which implements following functionality:".

And similar things to explain, find and fix an error, suggest an improvement, etc.  It is diverging from just helping with text, but seems to work in few tests I made. The major inconvenience for this use is that there is just one system prompt, so you need to update when working on different tasks. I need to check if I can employ language specific settings for vscode though, so for C/C++ I'd have different prompts than for markdown and latex.  It may be also possible to use file type from editor to configure system prompt dynamically to set the output language. Say, use some `${editor_language}`  as placeholder in prompt or something like that.

Btw, it is interesting, that if I have latex as input - gpt-4-turbo-preview results in latex as output. But if I don't have any special markup in the input,  but change a system prompt that it should generate latex markup  - I still get a markdown as output :). Seems I need to practice more with prompts to convince it.  Any way, I added few prompts like `"Write a few paragraphs on the following topic:"` and it greatly helps with content creation.

**Reply by ra-jeev** [reply]
Now I understand it. When i was adding the system prompt i thought of creating a list so that you can add many prompts, but then I thought that the list can be written in a single prompt itself.

Either way you're not asking for a list, you want to have different system prompts for different languages, right? This is interesting, I think I can give it some thought and add the support.

For code related stuff (fixing errors etc), you could try [Cursor](https://cursor.sh/). It is a port of VSCode with inbuilt support for this. (Haven't tried my extension there, need to verify if it works).

**Comment by vsukhoml** [comment]
Thanks for Cursor suggestion - it looks promising, though I found that autocompletion with Github's Copilot or other AI helpers is not what I need - I spend more time to changing generated code than to write on my own, but in some cases it is convenient to describe intended functionality and explicitly request its generation. And running AI helper in a separate window is no better than using your plugin, especially since common context for project can be part of prompt and include certain requirements like what to use / not to use, etc.

As to what can be enhanced my list of ideas:
1. Add delimiters to selected text. This is based on https://www.vellum.ai/blog/prompt-engineering-tips-to-boost-gpt-3-5-to-gpt-4-level and https://platform.openai.com/docs/guides/prompt-engineering/tactic-use-delimiters-to-clearly-indicate-distinct-parts-of-the-input --- so far I found that using ending like `following text:` works better for me.

2. Currently I can set different prompts for different workspaces which mostly address all my needs, but within a single workspace it can be better to either change system / user prompts based on language (say working on docs - text is better, when in code - code specific), and may be treat prompts as templates with few predefined variables like `${text}`, `${language}` and may be `${filename}`, though I don't know if it is useful for anything. I found that some AI plugins upload the whole file as a context for requests like `explain`, which seems to be overkill and significantly increase use of tokens. 

So, I'm not sure about best strategy to implement system prompts. I see more value in not just language specific system prompts, but in association of user prompt with specific system prompt. So, each user prompt refers to system prompt, and user prompts itself (those in context menu) would be the one depending on current language. Rationale: it may be beneficial to have different system prompts providing different roles for user prompts. 

If to implement it, then it is nice to enable configuration in which languages extension will be activated, so, if you want to add prompts for Javascript code gen/explanation/review, you can enable extension for Javascript (or may be not limit it at all, but generate context menu based on language).

**Reply by ra-jeev** [reply]
Thanks for such a detailed response @vsukhoml. Let me think about it how best we can incorporate your suggestions. On top of my head I'm thinking if snippets can play some role here. Other option can be a hover provider. Will need to check their usefulness.

**Comment by vsukhoml** [comment]
If to make system prompts referred from user prompts - it will require changes in code structure - `getOpenAIConfig` would need to get a system prompt as input, and it will be also argument to `handleAction` passed from `createAction`. Since it is likely to have fewer system prompts than user prompts, might be user prompt config can somehow refer to one or many system prompts, make some system prompt 'default', so no changes will be needed. Or turn config of system prompts into array and use by default first element if not requested differently in user prompt.

**Reply by ra-jeev** [reply]
> Add delimiters to selected text. I found that using ending like following text: works better for me.

I have been thinking about what you wrote. When the extension was using the legacy `v1/completions` endpoint (and when everything was hardcoded) it was using delimiters. And if you look at the prompts for each of the actions, it still uses them in a way. 

E.g, below is the structure for a rephrase action:

```json
{
  "title": "Rephrase the selected text",
  "description": "Rephrases the selected text",
  "prompt": "Rephrase the given text and make the sentences more clear and readable."
},
```

As you can see, we're using `given text`. The user is free to modify these prompts. Maybe by default we can add `Text: ` at the end of this.

I also checked the code and looks like I've a bug where the above prompt is being passed as a `system prompt` instead of being part of the `user prompt`. Here is the relevant code:

```ts
if (this.config.systemPrompt) {
  messages.push({ role: 'system', content: this.config.systemPrompt });
}

messages.push(
  { role: 'system', content: cmdPrompt },
  { role: 'user', content: text }
);
```

The `user prompt` should be: `{ role: 'user', content: cmdPrompt + text }` where `cmdPrompt` is the `prompt` part of each command. I'll fix it (and maybe add the `Text: ` delimiter by default to all the actions' prompts)

> may be treat prompts as templates with few predefined variables like ${text}, ${language} and may be ${filename}

I'm not so sure about this template thing. I mean what will we gain with this?

>  I see more value in not just language specific system prompts, but in association of user prompt with specific system prompt. So, each user prompt refers to system prompt, and user prompts itself (those in context menu) would be the one depending on current language.

This level of configuration might become complex, to make sense of, in the settings. Need to think how it be approached in a simple way.

As far as I've understood, what you want is (please do correct me):
1. Different `system prompts` (multiple) for different languages and a way to pick the desired system prompt during runtime
2. Different `Code Actions (Quick Fixes / Rewrites)` for each language.

Do let me know your thoughts.

Best,
Rajeev

**Reply by ra-jeev** [reply]
So I just tried that delimiter thing.

With the below system prompt:
> You are a helpful assistant. Your job is to perform the tasks related to rewriting text inputs given by the user. If the input text contains any special syntax then strictly follow that syntax, e.g. for markdown return markdown, for latex return latex etc. Do not return markdown for latex, and vice versa.

And the below command prompt for rephrase action:
> Rephrase the given text and make the sentences more clear and readable.\n\nText: 

We get a response in the below format:
> Rewritten Text: <the_actual_response>

So we've the extra `Rewritten Text: ` prefix. Does this match your findings?

One way we can fix this is: add it beforehand in the user prompt. So the user prompt becomes `<prompt>. Text: <the_text>. Your response:` But in this case `Your response:` will be hardcoded and is not flexible.  

Another way could be: suffix the action prompt to the system prompt itself (instead of making it a prefix of the user prompt) and let the user part be just the selected text without any delimiter. Not sure about this as well.

Your thoughts?

**Comment by vsukhoml** [comment]
> I also checked the code and looks like I've a bug where the above prompt is being passed as a system prompt

This is very interesting. I really didn't check what is the practical difference between system and user prompts, but this bug may be actually a feature. It just implies slightly different way to set user text with semicolon not needed and `given text` is more than enough. It also means `user prompt` is (or can be) in  the selected text - this makes it rather flexible. 

> I'm not so sure about this template thing. I mean what will we gain with this?

Templates are only useful for the case of code generation, where you'd instructing model to make output in specific language, which is current language for document, like C or JavaScript, as in some cases syntaxes are very close and hard to distinguish. It may be overkill for text-focused generation, and may be even better be done as a different product to not defocus key features of your extension. I just found that I prefer different method for AI-assisting than Copilot and similar and connected it to possible use of your extension, but developing this idea forward would probably require more efforts to support it. In the nutshell this is about mechanics of output - dialogue in separate window (or even switching apps) vs. input & output in same text window which saves time on context switching and reduce sudden distractions. And I don't like unpredictable, but pretending to be smart autocompletions.

> The user prompt should be: { role: 'user', content: cmdPrompt + text } where cmdPrompt is the prompt part of each command. I'll fix it (and maybe add the Text:  delimiter by default to all the actions' prompts)

I'm not sure anymore in this. The current behavior suits my needs, prompts do work.  Changing it to "proper" use may help or may make things worse. I'd make it a `switchable feature` - a flag to set `user` prompt as `user`  or `system`. Practically you can make a system prompt which doesn't need any `user`  prompt at all, treating it as a `given text`.

> As far as I've understood, what you want is (please do correct me):
>Different system prompts (multiple) for different languages and a way to pick the desired system prompt during runtime

Yes, but as you've discovered it is already here. I can just right `user prompts` in a way that they complement the `common system prompt`. And I practically can achieve what I want by making `common prompt` an empty in the extreme, putting all instructions into current `user` prompt going to be actually `system`. 

What is to be tested - language specific configuration in vscode. Like
```
 "[python]": {
 "writeAssistAi.systemPrompt": "You are an assistant to software engineer writing code in Python",
    },
```
Practically no new feature is needed  - i moved all project specific stuff into workspace's configuration and it works, and each workspace have its own set of prompts.

> Different Code Actions (Quick Fixes / Rewrites) for each language.

I didn't play with this really yet. Can be also solved by defining configs for workspaces or even folders. I  can merely change a system prompt setting proper context for that folder.

**Comment by vsukhoml** [comment]
> So we've the extra Rewritten Text:  prefix. Does this match your findings

No, this is different from what I meant. I don't want delimiters in the output. The idea was to only clearly indicate on the input side where text starts. But due to this "bug" where user prompt became system it is actually not needed. I really like the current behavior when I accustomed via trial and error my prompts to this behavior. Again, I don't really know how it will change the behavior if moved from system to user.

The kind of prompts I use are:
```
"prompt": "Explain following topic in professional tone, detailed and comprehensive:"

 "prompt": "Enhance the following text in professional, but casual tone and make the sentences more clear, readable, and concise:"
```
And current output doesn't include extra prefixes.

**Reply by vsukhoml** [reply]
And this is before I learned that user prompt is system.
I'm going to play with a different prompt like:
```
"prompt": "Explain topic given by user in professional tone, detailed and comprehensive."
```

**Reply by ra-jeev** [reply]
Ok. So, the new system prompt can be: `user defined system prompt (with some default value which i mentioned earlier) + the command specific prompt`. And the user prompt can be just the selected text. 

This was the case earlier, now the change is, instead of two system prompts, merge the two and create a single system prompt.

Though the issue i mentioned (the extra `Rewritten Text:` in API response) goes away if I add this: `You're just supposed to return the final output.` to the system prompt.

**Reply by vsukhoml** [reply]
> instead of two system prompts, merge the two and create a single system prompt.

I hope there is no difference in behavior compared to sending different system prompts twice. Changes in any are actually applied, so guess OpenAI merges them internally. But this doesn't seem to be a well documented behavior, however example at https://platform.openai.com/docs/guides/text-generation/chat-completions-api shows two user prompts. 

I tested that prompts like the below going into "system" works fine:
```
"prompt": "Enhance user text in professional, but casual tone and make the sentences more clear, readable, and concise"
```

**Reply by ra-jeev** [reply]
Multiple user prompts are fine. As this is a chat model, so there can be multiple user messages, and the corresponding assistant messages.

**Reply by ra-jeev** [reply]
Finally made the action prompt a part of user prompt itself using the below syntax:

```js
const userPrompt = `${cmdPrompt}${
  cmdPrompt.endsWith(':') ? '\n\n' : ':\n\n'
}${text}`;
messages.push({ role: 'user', content: userPrompt });
```

This seems to work fine. Have made a minor release as well with this.

**Reply by vsukhoml** [reply]
Ok. I'd still add a flag to control wherever action prompt should be a user prompt or a system prompt - I can't find much details on how it actually impact output beyond some general comments on what is "better" to put where. 

https://community.openai.com/t/what-is-the-difference-between-putting-the-ai-personality-in-system-content-and-in-user-content/194938/4 suggests that putting system prompt last can change a behavior. So, interplay between settings on real life output is yet to be explored ❓

**Reply by ra-jeev** [reply]
Ok. In my opinion we shouldn't have any such issue as we always have one system prompt and one user prompt, never more that that. We're never trying to create a chat thread.

**Reply by ra-jeev** [reply]
Now I'm thinking about the system prompts array, how best to integrate such a thing.

**Comment by vsukhoml** [comment]
I tried 0.3.1 and it seems to work in a same way as 0.3.0 (meaning everything still works with same prompts, which is good).

> In my opinion we shouldn't have any such issue as we always have one system prompt and one user prompt

Yes, it seems this is for chat use case only.

> Now I'm thinking about the system prompts array, how best to integrate such a thing.

May be additional optional parameter to action in addition to `prompt` - `system_prompt`? So you can specify it if it is beneficial for some use case. 

I'm still trying to understand what is practically changing when you have move prompt from user to system or in reverse. 
https://community.openai.com/t/the-system-role-how-it-influences-the-chat-behavior/87353/34, https://community.openai.com/t/understanding-role-management-in-openais-api-two-methods-compared/253289

It seems that major changes are for the chat use case. So far I use `system` to add some additional context about problem domain which is common for all requests. But I guess it will work roughly the same way if all comes to `user`. Will play in the playground how it works. I suspect system/user here are very similar to "delimiters", but this needs to be tested.

**Reply by ra-jeev** [reply]
@vsukhoml Haven't been able to find time to work on this. Been traveling and busy with other stuffs. Maybe in a couple of weeks I should be able to find some time :-).

**Reply by vsukhoml** [reply]
@ra-jeev no worries! so far existing functionality is great! So far I see that there is no big difference with system vs. user prompts, and don't see much impact of delimiters. Seems that all those tricks are very use case dependent.

**Reply by ra-jeev** [reply]
@vsukhoml Finally I was able to find some time and have made a new release (`v0.4.0`) just now. This release has the following changes:
1. Dynamic OpenAI config change reflection (earlier if you change the model or say system prompt or temperature etc, it was not reflecting immediately).
2. Have made `System Prompt`, `QuickFixes` & `RewriteOptions` language overridable. So now we can have language specific system prompts and actions.

Please do share your feedback on the changes when you find time and are able to use the new release. I hope this is what you had in mind when you wanted multiple system prompts and language specific config.

If everything goes well, then maybe I can open up this extension for all languages so that one can setup their own actions for mundane coding tasks as well. 

 Thanks

**Reply by vsukhoml** [reply]
I justed tested 0.4.0. Everything that worked before, still works :-)

For some reason I couldn't make language specific options to work though - tried for `"[markdown]": {}`, and `"[latex]": {}`, but got:
`This setting cannot be applied because it is not registered as language override setting.`, despite I restarted extension.

But when I restarted `vscode` it seems to start work. However, I got a different effect - if I have say `writeAssistAi.quickFixes` redefined for language, `writeAssistAi.rewriteOptions` from common is not used. Same for `writeAssistAi.rewriteOptions`, which happened to be a documented behavior I missed at first glance.

So, it seems everything works as expected!

I switched to `gpt-4-turbo-preview` which seems to produce better results for me and from time to time even take into account that I want Latex to be generated (when Latex mark-up is in, it works almost always, but it continues to output in markdown for requests w/o markup in most of times).

So far it seems very flexible and easily configurable and I like it a lot - saves my time with wordsmithing!

**Comment by ra-jeev** [comment]
> For some reason I couldn't make language specific options to work though - tried for `"[markdown]": {}`, and `"[latex]": {}`, but got:
> `This setting cannot be applied because it is not registered as language override setting.`, despite I restarted extension.

interesting. IMO it should have worked on update. Will check if I can reproduce this.

> But when I restarted `vscode` it seems to start work. However, I got a different effect - if I have say `writeAssistAi.quickFixes` redefined for language, `writeAssistAi.rewriteOptions` from common is not used. Same for `writeAssistAi.rewriteOptions`, which happened to be a documented behavior I missed at first glance.

Yes, it is either take all or nothing from the default actions. This can be changed if you feel that for a unconfigured action key (`quickFixes` or `rewriteOptions`) the default actions for that key should take over. Or, if the documentation needs to be changed to reflect it better, then I can do that as well. Please let me know what are your thoughts on this.

> So far it seems very flexible and easily configurable and I like it a lot - saves my time with wordsmithing!

I'm glad to hear this :-)
